src/library/Cardano/Fee.hs

```diff
@@ -417,35 +416,32 @@ coalesceDust (DustThreshold threshold) coins =
 
 -- | Computes how much is left to pay given a particular selection
 remainingFee
-    :: HasCallStack
-    => Buildable i
-    => Buildable o
-    => FeeOptions i o
+    :: (HasCallStack, Buildable i, Buildable o)
+    => FeeEstimator i o
     -> CoinSelection i o
     -> Fee
-remainingFee opts s = do
-    if fee >= diff
-    then Fee (fee - diff)
-    else do
+remainingFee FeeEstimator {estimateFee} s
+    | fee >= diff =
+        Fee (fee - diff)
+    | feeDangling >= diff =
+        Fee (feeDangling - fee)
+    | otherwise =
         -- NOTE
         -- The only case where we may end up with an unbalanced transaction is
         -- when we have a dangling change output (i.e. adding it costs too much
         -- and we can't afford it, but not having it result in too many coins
         -- left for fees).
-        let Fee feeDangling =
-                estimateFee opts $ s { change = [Coin (diff - fee)] }
-        if (feeDangling >= diff)
-            then Fee (feeDangling - fee)
-            else error $ unwords
-                [ "Generated an unbalanced tx! Too much left for fees"
-                , ": fee (raw) =", show fee
-                , ": fee (dangling) =", show feeDangling
-                , ", diff =", show diff
-                , "\nselection =", pretty s
-                ]
+        error $ unwords
```

Comment:
would be good to explicitly test it, ie., that this error is emitted

Reply:
> would be good to explicitly test it, ie., that this error is emitted

Definitely! That is actually the subject of PR https://github.com/input-output-hk/cardano-coin-selection/pull/36 (still WIP).